File: solutions/javascript/matching-brackets/1/matching-brackets.js
//
// matching brackets
/* isPaired returns true or false depending on whetcher brackets, braces or parenthesis are pair in the expression passed as an argument*/

const matches = (left, right) => {
    return (left=='(' && right==')') || 
          (left=='[' && right==']') ||
          (left=='{' && right=='}');
}

export const isPaired = (expression) => {
  let lifo = [];
  for(let symbol of expression){
    if(symbol=='('||symbol=='['||symbol=='{')
      lifo.push(symbol);
    if(symbol==')'||symbol==']'||symbol=='}'){
       if(0==lifo.length)
         return false;
       if (!matches(lifo.pop(),symbol))
         return false;      
    }    
  }
  return (0==lifo.length);    
};

File: solutions/dart/matching-brackets/1/lib/matching_brackets.dart
class MatchingBrackets {
  bool isPaired(String text){
    List<String> brackets = [];
    for(var s in text.split('')){
      if(s == '(' || s == '{' || s=='[')
        brackets.add(s);
      else if (s== ')' || s==']' || s=='}'){
        if(brackets.length>0 && matches(brackets.last,s)){
          brackets.removeLast();
        } else 
          return false;
      }             
    }
    return brackets.length==0;
  }

  bool matches(String l, String r){
    return (l=='(' && r == ')') || (l=='[' && r == ']')
      || (l=='{' && r == '}');
  }
}
